src/vs/workbench/api/node/mainThreadHandlerRegistry.ts

```diff
@@ -0,0 +1,57 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+/**
+ * MainThreadHandlerRegistry maintains a collection of objects by handle.
+ */
```

Comment:
I wish I could whiteboard this for you, but `Main Thread <-> Extension Host` communication is not in the same process, so requests and responses are serialized (to JSON) over IPC. Functions (callbacks) can't be serialized over the wire :) so the Main thread has to
1. Save the callback in a dictionary with a new handle
2. Send the handle to the extension
3. The extension sends the main thread a message with that handle when it wants to report progress
4. The main thread looks up the callback when it receives a message with the corresponding progress handle and calls the callback with the data

Hopefully that makes sense.